criu/mem.c

```diff
@@ -172,17 +175,58 @@ static int generate_iovs(struct pstree_item *item, struct vma_area *vma, struct
 	unsigned long pages[3] = {};
 	int ret = 0;
 
+	static char *zero_page = NULL;
+	static char *remote_page = NULL;
+	int zero = 0;
+	struct iovec  local[2];
+	struct iovec  remote[1];
+	int nread = 0;
+	if (opts.skip_zero_pages && zero_page == NULL) {
+		zero_page = xmalloc(PAGE_SIZE);
+		remote_page = xmalloc(PAGE_SIZE);
+		if (zero_page == NULL || remote_page == NULL) {
+			pr_warn("Can't allocate memory - disabling --skip-zero-pages\n");
+			opts.skip_zero_pages = 0;
+		} else	{
+			memzero(zero_page, PAGE_SIZE);
+			local[0].iov_base = remote_page;
+			local[0].iov_len = PAGE_SIZE;
+			remote[0].iov_base = (void *) 0x0;
+			remote[0].iov_len = PAGE_SIZE;
+		}
+	}
+
 	nr_to_scan = (vma_area_len(vma) - *off) / PAGE_SIZE;
 
 	for (pfn = 0; pfn < nr_to_scan; pfn++) {
 		unsigned long vaddr;
 		unsigned int ppb_flags = 0;
 		int st;
 
-		if (!should_dump_page(vma->e, at[pfn]))
-			continue;
-
 		vaddr = vma->e->start + *off + pfn * PAGE_SIZE;
+		/*
+		 * If should_dump_page() returns true, it means the page is in the dumpees resident memory
+		 * (i.e. bit 63 of the page frame number 'at[pfn]' is set) but it is not the zero-page.
+		 */
+		if (should_dump_page(vma->e, at[pfn])) {
+			if (opts.skip_zero_pages) {
+				remote[0].iov_base = (void*)vaddr;
+				nread = process_vm_readv(item->pid->real, local, 1, remote, 1, 0);
```

Comment:
I don't like the idea to read process memory twice. We have to avoid this.

btw: https://github.com/checkpoint-restore/criu/pull/2292 solves the same problem in a more optimal way.

Reply:
I agree that it would be better if we can avoid reading process memory twice.

> https://github.com/checkpoint-restore/criu/pull/2292 solves the same problem in a more optimal way.

I think this is different. In https://github.com/checkpoint-restore/criu/pull/2292, we exclude pages with zero PFN (`PAGE_IS_PFNZERO`), while this option skips zero-filled memory (e.g., memory that has been filled with zeros using `memset()` would be skipped with this option).

Reply:
Is this just a performance or a correctness issue? If it's just about performance, I think the benefit might justify the additional overhead and after all the feature is on by default.

In the case this is a correctness problem, do you have a suggestion how we can avoid this?

PS: and yes, @rst0git is right - this change is about skipping regular pages  which are filled with only zero bytes.

Reply:
> This can be useful for runtimes like Java, which often allocate large memory regions without fully using them (e.g. for the heap). For a simple Helloworld Java program, this new feature shrinks the image size be about 20% from 13mb to 11mb.

> we exclude pages with zero PFN (PAGE_IS_PFNZERO), while this option skips zero-filled memory

I believe we need a zdtm test, which can reproduce such a zero page without PAGE_IS_PFNZERO but with zero data. Maybe, we even want to fix kernel to report such a page as "PAGE_IS_PFNZERO" instead.

I agree with Andrei that manually checking page content with memcmp is an anti-pattern. Nack from me, at list in current state.

Reply:
I also agree. Reading a page second time and using `memcmp()` sounds not optimal.

I still like the idea of not including empty pages in the checkpoint, but it sounds difficult.

If the kernel could track it, that would be nice. Not sure the kernel has a better alternative than `memcmp()` to find a zeroed memory page.

At this point I think it would be nice to see some numbers. How much faster is restoring if something like this PR is applied. Although I don't like the `memcmp()` it would only be used if the corresponding command-line option is explicitly selected by the user. Maybe that makes it acceptable. Can the second reading of the page be avoided?

Maybe some post-processing of the checkpoint image would be an alternative. Remove the empty pages after checkpointing and have support during restore to handle pages like this.

Reply:
> I believe we need a zdtm test, which can reproduce such a zero page without PAGE_IS_PFNZERO but with zero data.

The code snippet in https://github.com/checkpoint-restore/criu/pull/2331#discussion_r1459073455 can be used to reproduce it:
- `criu dump` creates `pages-1.img` with size 4.1GB
-  when `--skip-zero-pages` is used, `pages-1.img` should be 2.1GB.

> Maybe, we even want to fix kernel to report such a page as "PAGE_IS_PFNZERO" instead.

If understand correctly, page frame number (PFN) corresponds to the physical address of the page. If a process writes zeros to a memory page, the PFN for that page would not be zero.

Reply:
> > This can be useful for runtimes like Java, which often allocate large memory regions without fully using them (e.g. for the heap). For a simple Helloworld Java program, this new feature shrinks the image size be about 20% from 13mb to 11mb.
> 
> > we exclude pages with zero PFN (PAGE_IS_PFNZERO), while this option skips zero-filled memory
> 
> I believe we need a zdtm test, which can reproduce such a zero page without PAGE_IS_PFNZERO but with zero data. Maybe, we even want to fix kernel to report such a page as "PAGE_IS_PFNZERO" instead.
> 

The kernel obviously can not do this, at least not until we run with [KSM and `use_zero_pages`](https://docs.kernel.org/admin-guide/mm/ksm.html) which comes with its own set of problems and overhead.

> I agree with Andrei that manually checking page content with memcmp is an anti-pattern. Nack from me, at list in current state.

That's actually exactly what this change is about - spending extra time during snapshot in order to decrease the resulting image size. It is basically KSM for pages containing only zero byte. I don't see how we could do that without reading the content of the page first. And again, as called out by @adrianreber, this is an optional feature which is guarded by its own command line option. So it will not impact anybody who is not using the new `--skip-zero-pages` option.

Reply:
> The code snippet in https://github.com/checkpoint-restore/criu/pull/2331#discussion_r1459073455 can be used to reproduce it:

> The kernel obviously can not do this, at least not until we run with [KSM](https://docs.kernel.org/admin-guide/mm/ksm.html)

Ok, I see that in the mentioned example we have this problem. But that is because the example's code is written sub-optimally, isn't it?

If we slightly change your example like this:
```
diff --git a/malloc-test.c b/malloc-test.c
index b959134..75d13ec 100644
--- a/malloc-test.c
+++ b/malloc-test.c
@@ -6,7 +6,7 @@
 int main() {
        /* Allocate 4GB of memory */
        size_t size = 4ULL * 1024 * 1024 * 1024;
-       char *memory = (char *)malloc(size);
+       char *memory = (char *)calloc(1, size);
 
        if (memory == NULL) {
                fprintf(stderr, "Memory allocation failed\n");
@@ -15,7 +15,6 @@ int main() {
 
        /* Fill the first half with 1 and the second half with 0 */
        memset(memory, 1, size / 2);
-       memset(memory + size / 2, 0, size / 2);
 
        while(1)
                sleep(1);
```
we will have the same result, but avoid zero-filled pages in RSS.

For instance (sorry I don't have free 4Gb on my laptop now so I changed allocation size to 2Gb, but it should not matter), in your example I get this picture in smaps:
```
7f9983374000-7f9a03377000 rw-p 00000000 00:00 0 
Size:            2097164 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
Rss:             2097160 kB
```
if we change to calloc it saves us half of the memory:
```
7f565d25e000-7f56dd261000 rw-p 00000000 00:00 0
Size:            2097164 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
Rss:             1048584 kB
```

And if I dump the latter I get:
```
du -h /tmp/images-dir/pages-1.img 
1.1G	/tmp/images-dir/pages-1.img
```

So here is next question: Do we really need to cover malloc+memset(0) case?

I'm not sure here, but maybe...

Reply:
This is certainly not useful for everyone (and that's why it is a command line option which is off by default), but as I already wrote in the initial PR:

> *This can be useful for runtimes like Java, which often allocate large memory regions without fully using them (e.g. for the heap, code cache, metaspace). For a simple Helloworld Java program, this new feature shrinks the image size be about 20% from 13mb to 11mb.*

I don't think we should mandate specific programming styles in order for applications to work well with CRIU but instead try to optimize them all.

Reply:
It's not just the "malloc+memset(0)" use case. Java does [mmap and pretouch memory](https://www.ibm.com/docs/en/semeru-runtime-ce-z/17?topic=options-xx-alwayspretouch) so we can have a lot of pages with only zero content (but not the kernel zero page) in some scenarios.

Reply:
> Is this just a performance or a correctness issue? If it's just about performance, I think the benefit might justify the additional overhead and after all the feature is on by default.

I don't understand why we need to read process pages to do this check? Why can't we do that before dumping this pages into the image (page_xfer_dump_pages)?

Reply:
> I believe memset(0) is used in specific scenarios where applications/libraries are dealing with sensitive data and/or use custom memory management.

> It's not just the "malloc+memset(0)" use case. Java does mmap and pretouch memory so we can have a lot of pages with only zero content (but not the kernel zero page) in some scenarios.

1. First,
<details><summary>in current version of code it does not seem to work on my system:</summary>
<p>

I use the example dumpee code from above comments (2G instead of 4G):
```
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

int main() {
	/* Allocate 2GB of memory */
	size_t size = 2ULL * 1024 * 1024 * 1024;
	char *memory = (char *)malloc(size);

	if (memory == NULL) {
		fprintf(stderr, "Memory allocation failed\n");
		return 1;
	}

	/* Fill the first half with 1 and the second half with 0 */
	memset(memory, 1, size / 2);
	memset(memory + size / 2, 0, size / 2);

	printf("Enter any char to stop\n");
	scanf("%*c");

	free(memory);
	return 0;
}
-------------------------------------------------
[root@turmoil tmp]# gcc -o malloc-test malloc-test.c
[root@turmoil tmp]# ./malloc-test 
Enter any char to stop
```
Here is what smaps show after I run the above:
```
[root@turmoil snorch]# grep 2097164 -A3 -B1 /proc/$(pidof malloc-test)/smaps
7fd2fbee9000-7fd37beec000 rw-p 00000000 00:00 0 
Size:            2097164 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
Rss:             2097160 kB
```
Let's dump it:
```
[root@turmoil tmp]# /home/snorch/devel/ms/criu/criu/criu dump -v4 -o dump.log -t $(pidof malloc-test) -j -D /images-dir/
[root@turmoil tmp]# du -h /images-dir/pages-1.img 
2.1G	/images-dir/pages-1.img
```

CRIU is at b23ad2242 ("Don't dump pages which only contain zero bytes").

</p>
</details>

Second, If Java put so much effort to have those zeroed pages in RSS isn't it a bad idea to restore those pages like they are "PAGE_IS_PFNZERO" ones? =)

Reply:
> Second, If Java put so much effort to have those zeroed pages in RSS isn't it a bad idea to restore those pages like they are "PAGE_IS_PFNZERO" ones? =)

@Snorch, Java (or OpenJDK based JVMs to be more specific) was not designed and optimized for cloud/container use cases but rather for large, monolithic application servers. In the old days, pretouching/zeroing memory was a way to pre-allocate physical memory and not potentially get it from swap later. For current cloud/container use cases the huge memory footprint can be problem. With check-pointing, a small image is more important and COWing a PAGE_IS_PFNZERO page is much faster then loading and populating it from disk.

Reply:
> I don't understand why we need to read process pages to do this check? Why can't we do that before dumping these pages into the image (page_xfer_dump_pages)?

Thanks a lot for your suggestion @avagin. I'm short of time for the next week because of FOSDEM/Jfokus but I'll try to come up with a new version which moves the zero check to `page_xfer_dump_pages()` afterwards.

Reply:
@simonis FOSDEM was my favorite conference when I lived close by. btw @mihalicyn are there too, he is one of criu maintainers. He will be happy to help with any questions.